windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -0,0 +1,1348 @@
+        begin
+          Result := False;
+          Exit;
+        end;
+        // Else Time to check for updates
+        Result := True;
+      end;
+    finally
+      RegistryErrorControlled.Free;
+    end;
+  except
+    { we will not run the check if an error occurs reading the settings }
+    on E:Exception do
+    begin
+      Result := False;
+      FErrorMessage := E.Message;
+      Exit;
+    end;
+  end;
+end;
+
+function TUpdateStateMachine.GetState: TStateClass;
+begin
+  Result := TStateClass(CurrentState.ClassType);
+end;
+
+procedure TUpdateStateMachine.SetState(const Value: TStateClass);
+begin
+  if Assigned(CurrentState) then
+  begin
+    CurrentState.Exit;
+  end;
+
+  SetStateOnly(Value);
+
+  if Assigned(CurrentState) then
+  begin
+    CurrentState.Enter;
+  end
+  else
+  begin
+   // TODO: #10210 Error log for Unable to set state for Value
+  end;
+
+end;
+
+procedure TUpdateStateMachine.SetStateOnly(const Value: TStateClass);
+begin
+  if Value = IdleState then
+  begin
+    CurrentState := FIdle;
+  end
+  else if Value = UpdateAvailableState then
+  begin
+    CurrentState := FUpdateAvailable;
+  end
+  else if Value = DownloadingState then
+  begin
+    CurrentState := FDownloading;
+  end
+  else if Value = WaitingRestartState then
+  begin
+    CurrentState := FWaitingRestart;
+  end
+  else if Value = InstallingState then
+  begin
+    CurrentState := FInstalling;
+  end
+  else if Value = RetryState then
+  begin
+    CurrentState := FRetry;
+  end
+  else if Value = PostInstallState then
+  begin
+    CurrentState := FPostInstall;
+  end;
+end;
+
+function TUpdateStateMachine.ConvertEnumState(const TEnumState: TUpdateState) : TStateClass;
+begin
+  case  TEnumState of
+    usIdle: Result := IdleState;
+    usUpdateAvailable: Result := UpdateAvailableState;
+    usDownloading: Result := DownloadingState;
+    usWaitingRestart: Result := WaitingRestartState;
+    usInstalling: Result := InstallingState;
+    usRetry: Result := RetryState;
+    usPostInstall: Result := PostInstallState;
+  else
+    // TODO: #10210 Log error unknown state setting to idle
+    Result := IdleState;
+  end;
+end;
+
+procedure TUpdateStateMachine.HandleCheck;
+begin
+  CurrentState.HandleCheck;
+end;
+
+function TUpdateStateMachine.HandleKmShell;
+begin
+  Result := CurrentState.HandleKmShell;
+end;
+
+procedure TUpdateStateMachine.HandleDownload;
+begin
+  CurrentState.HandleDownload;
+end;
+
+procedure TUpdateStateMachine.HandleAbort;
+begin
+  CurrentState.HandleAbort;
+end;
+
+procedure TUpdateStateMachine.HandleInstallNow;
+begin
+  CurrentState.HandleInstallNow;
+end;
+
+function TUpdateStateMachine.CurrentStateName: string;
+begin
+  Result := CurrentState.StateName;
+end;
+
+{ State Class Memebers }
+constructor TState.Create(Context: TUpdateStateMachine);
+begin
+  bucStateContext := Context;
```

Comment:
```suggestion
  inherited Create;
  bucStateContext := Context;
```

Reply:
TState is an abstract base clase, does it need the inherited Create?

Reply:
The reason I added it, is if we add a Create constructor to the base class in the future, we won't accidentally skip it in child classes.

Reply:
That would make sense if this was a child but this is the base class constructer.

Reply:
Yeah okay, nbd, but even then I have traditionally added an inherited constructor just in case the class def changes in the future and that way it doesn't get forgotten (because this has bitten me in the past).